File: leetcode-cn/1222.0_Queens_That_Can_Attack_the_King.py
'''
44 ms
15.8 MB
'''
class Solution:
    def queensAttacktheKing(self, queens: List[List[int]], king: List[int]) -> List[List[int]]:
        dirs = [[-1, 0], [1, 0], [0, -1], [0, 1], [-1, 1], [-1, -1], [1, -1], [1 ,1]]
        ans = []
        for i in range(8):
            x, y = king 
            while True:
                x += dirs[i][0]
                y += dirs[i][1]
                if x < 0 or x >= 8 or y < 0 or y >= 8:
                    break
                if [x, y] in queens:
                    ans.append([x, y])
                    break             
        return ans 
        

File: js/1222.0_Queens_That_Can_Attack_the_King.js
/*
68 ms
43.4 MB
*/
const directions = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]

/**
 * @param {number[][]} queens
 * @param {number[]} king
 * @return {number[][]}
 */
var queensAttacktheKing = function(queens, king) {
    const isQueen = new Array(8).fill(null).map(() => new Array(8).fill(false));
    for (const [x, y] of queens) {
        isQueen[x][y] = true;
    }
    var ans = [];
    for (const [dx, dy] of directions) {
        let x = king[0] + dx;
        let y = king[1] + dy;
        while (0 <= x && x < 8 && 0 <= y && y < 8) {
            if (isQueen[x][y]) {
                ans.push([x, y]);
                break;
            }
            x += dx;
            y += dy;
        }
    }
    return ans;
};
